server/src/main/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricAggregator.java

```diff
@@ -23,38 +23,66 @@
 import org.apache.lucene.search.Scorable;
 import org.apache.lucene.search.ScoreMode;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.lease.Releasables;
 import org.elasticsearch.common.util.CollectionUtils;
+import org.elasticsearch.common.util.ObjectArray;
 import org.elasticsearch.script.Script;
 import org.elasticsearch.script.ScriptedMetricAggContexts;
+import org.elasticsearch.script.ScriptedMetricAggContexts.MapScript;
 import org.elasticsearch.search.aggregations.Aggregator;
 import org.elasticsearch.search.aggregations.InternalAggregation;
 import org.elasticsearch.search.aggregations.LeafBucketCollector;
 import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
 import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.search.lookup.SearchLookup;
 
 import java.io.IOException;
+import java.util.HashMap;
 import java.util.Map;
 
 class ScriptedMetricAggregator extends MetricsAggregator {
+    /**
+     * Estimated cost to maintain a bucket. Since this aggregator uses
+     * untracked java collections for its state it is going to both be
+     * much "heavier" than a normal metric aggregator and not going to be
+     * tracked by the circuit breakers properly. This is sad. So we pick a big
+     * number and estimate that each bucket costs that. It could be wildly
+     * inaccurate. We're sort of hoping that the real memory breaker saves
+     * us here. Or that folks just don't use the aggregation.
+     */
+    private static final long BUCKET_COST_ESTIMATE = 1024 * 5;
```

Comment:
what made you choose this number beyond just "make it large"?

Reply:
Drive-by comment: I wonder if it would make sense to expose this as a configurable option for advanced users?  (perhaps in a followup PR)

Reasoning being that we can't figure it out easily, it's already an advanced agg, so a user might actually be able to give us a reasonable estimate.  And if they don't set it, we fall back to Big Number.

Dunno, might be a terrible idea :)

Reply:
The number is the original default "weight" of an aggregator. So it is basically what we used to have. I didn't want to just set it to the default weight because it doesn't really have anything to do with it. Other than it is what we were using.

Making it configurable is certainly interesting! I don't folks are going to have a good idea of what to set it to without having done some java hacking though. We try not to leak stuff like that to our users. OTOH `scripted_metric` is pretty bonkers so if we are going to leak that sort of thing anywhere, here is it.

Reply:
Yeah I don't feel super great about the idea either.  But then again, we're essentially allowing a user to write a custom map-reduce job that runs on ES, so exposing this kind of accounting isn't too crazy in that context.

There's probably a case for _forcing_ the user to estimate it too, since I've definitely seen very bad scripts knock over clusters before (maps holding high cardinality IDs, etc).  I think that's probably a bridge too far, alas :)